### ISSUE TITLE
Customize / make more readable output of promise new promise (promise based dynamic remotes)

### ISSUE THREAD

**Issue by trevorgithub** [webpack/webpack#16475]
## Feature request

**What is the expected behavior?**
Promise new promise syntax ([promised based dynamic remotes](https://webpack.js.org/concepts/module-federation/#promise-based-dynamic-remotes)) gives excellent flexibility in defining remotes.  However, the build output of this is too verbose, showing the entire body of the function in the build output.

**What is motivation or use case for adding/changing the behavior?**
Promise based dynamic remotes give great flexibility.  But, the build output is hard to read, particularly when more than 1 promised based dynamic remote is specified. Other details of the build output can be obscured by the volume of the promised based dynamic remote output.

**How should this be implemented in your opinion?**
It looks like this output comes from the [readableIdentifier ](https://github.com/webpack/webpack/blob/main/lib/ExternalModule.js#L430)function in ExternalModule.js. readableIdentifier appears to be part of the Module class, and is implemented differently for different module types. For external modules, it looks like this:
```
         /**
	 * @param {RequestShortener} requestShortener the request shortener
	 * @returns {string} a user readable identifier of the module
	 */
	readableIdentifier(requestShortener) {
		return "external " + JSON.stringify(this.request);
	}
```

One possibly naive revision would be to truncate the stringified output to some maximum length.  For example, change ExternalModule.js' method to look like this:
```
         /**
	 * @param {RequestShortener} requestShortener the request shortener
	 * @returns {string} a user readable identifier of the module
	 */
	readableIdentifier(requestShortener) {
		const maxLength = 80;
		const stringifiedRequest = JSON.stringify(this.request);
		return "external " + (stringifiedRequest.length > maxLength ? stringifiedRequest.substring(0, maxLength) + '...': stringifiedRequest)
	}

```

**Are you willing to work on this yourself?**
yes, although someone else likely has a better understanding.  I'd be happy to test and otherwise assist

**Comment by alexander-akait** [comment]
> But, the build output is hard to read, particularly when more than 1 promised based dynamic remote is specified.

Why you need readable output here, debug? Can you provide more information?

**Comment by trevorgithub** [comment]
> > But, the build output is hard to read, particularly when more than 1 promised based dynamic remote is specified.
> 
> Why you need readable output here, debug? Can you provide more information?

Yes, for debugging or otherwise understanding the build results.  It's hard to follow what's happening in the build when you get overwhelmed with a very big block of text (in this case, the text of a function definition).  The text is unbounded, based on the definition of one's promise function definition.  Here's a partial example output, for a single use of promised based dynamic remotes:

```
remote app1/Component 6 bytes (remote) 6 bytes (share-init) [built] [code generated]
external "new Promise(resolve => {\n                    const urlParams = new URLSearchParams(window.location.search)\n                    const version = urlParams.get('app1VersionParam')\n
// This part depends on how you plan on hosting and versioning your federated modules\n                    const remoteUrlWithVersion = 'http://localhost:3001/' + version + '/remoteEntry.js'\n
  const script = document.createElement('script')\n                    script.src = remoteUrlWithVersion\n                    script.onload = () => {\n                      // the injected script has loaded and
 is available on window\n                      // we can now resolve this Promise\n                      const proxy = {\n                        get: (request) => window.app1.get(request),\n
     init: (arg) => {\n                          try {\n                            return window.app1.init(arg)\n                          } catch(e) {\n                            console.log('remote containe
r already initialized')\n                          }\n                        }\n                      }\n                      resolve(proxy)\n                    }\n                    // inject this script w
ith the src set to the versioned remoteEntry.js\n                    document.head.appendChild(script);\n                  })\n                  " 42 bytes [built] [code generated]

webpack 5.72.1 compiled with 1 warning in 23906 ms
Done in 27.52s.
```

The above (partial) output is based on the sample use of a dynamic function in the docs (repeated below):

```
module.exports = {
  plugins: [
    new ModuleFederationPlugin({
      name: 'host',
      remotes: {
        app1: `promise new Promise(resolve => {
      const urlParams = new URLSearchParams(window.location.search)
      const version = urlParams.get('app1VersionParam')
      // This part depends on how you plan on hosting and versioning your federated modules
      const remoteUrlWithVersion = 'http://localhost:3001/' + version + '/remoteEntry.js'
      const script = document.createElement('script')
      script.src = remoteUrlWithVersion
      script.onload = () => {
        // the injected script has loaded and is available on window
        // we can now resolve this Promise
        const proxy = {
          get: (request) => window.app1.get(request),
          init: (arg) => {
            try {
              return window.app1.init(arg)
            } catch(e) {
              console.log('remote container already initialized')
            }
          }
        }
        resolve(proxy)
      }
      // inject this script with the src set to the versioned remoteEntry.js
      document.head.appendChild(script);
    })
    `,
      },
      // ...
    }),
  ],
};
```
As noted, this is a single use of such a function.  Every remote that has a similar promised based dynamic remote definition will add to the output length.  And this sample dynamic remote is simpler than what I'm actually doing; my output is considerably longer for a single function.

**Comment by alexander-akait** [comment]
Yeah, I see, feel free to send a PR

**Comment by alexander-akait** [comment]
One note we shoul do it here https://github.com/webpack/webpack/blob/main/lib/stats/DefaultStatsFactoryPlugin.js, because it is only for stats

**Comment by trevorgithub** [comment]
> One note we shoul do it here https://github.com/webpack/webpack/blob/main/lib/stats/DefaultStatsFactoryPlugin.js, because it is only for stats

Thanks Alexander.  I could definitely be mistaken, but I think the verbose output I'm seeing is coming from [readableIdentifier](https://github.com/webpack/webpack/blob/main/lib/ExternalModule.js#L430) method in ExternalModule.js.  When I changed my local copy of that file, I was able to get the desired output.  If there's an alternate way to accomplish the same goal in DefaultSatsFactoryPlug.js, I'm definitely open to changing it there instead.  But I think I'll need a little bit of guidance :)

This is the partial call stack I see when I put a breakpoint in readableIdentifer, if it helps:

```
readableIdentifier (ExternalModule.js:430)
(anonymous) (ModuleInfoHeaderPlugin.js:191)
(anonymous) (VM3189:7)
(anonymous) (JavascriptModulesPlugin.js:596)
tryRunOrWebpackError (HookWebpackError.js:83)
renderModule (JavascriptModulesPlugin.js:594)
(anonymous) (JavascriptModulesPlugin.js:743)
(anonymous) (Template.js:303)
renderChunkModules (Template.js:300)
renderMain (JavascriptModulesPlugin.js:738)
render (JavascriptModulesPlugin.js:271)
(anonymous) (Compilation.js:4624)
(anonymous) (Cache.js:91)
(anonymous) (MemoryCachePlugin.js:45)
(anonymous) (Cache.js:91)
(anonymous) (VM1752:19)
get (Cache.js:75)
get (CacheFacade.js:111)
(anonymous) (Compilation.js:4560)
arrayEach (async.js:2405)
each (async.js:2846)
(anonymous) (Compilation.js:4549)
symbolIterator (async.js:3482)
timesSync (async.js:2297)
eachLimit (async.js:3463)
createChunkAssets (Compilation.js:4526)
(anonymous) (Compilation.js:3100)
(anonymous) (Compilation.js:3277)
(anonymous) (async.js:2818)
done (async.js:3522)
(anonymous) (Compilation.js:3240)
(anonymous) (Compilation.js:3349)
(anonymous) (HookWebpackError.js:68)
(anonymous) (VM1770:15)
store (Cache.js:107)
<lots more stuff below>
```

**Comment by alexander-akait** [comment]
@trevorgithub `readableIdentifer` is used in more methods, it is not safe to do it inside `readableIdentifer`, just create a helper function, for example - `wrapReadableIdentifier` (or any other nice name) with your logic and use it for each `readableIdentifier` call in https://github.com/webpack/webpack/blob/main/lib/stats/DefaultStatsFactoryPlugin.js

**Comment by trevorgithub** [comment]
Yes, certaintly readableIdentifier appears to be used in a lot of places.

I think what you're suggesting is something like this to be added in DefaultStatsFactoryPlugin.js:
```
const wrapReadableIdentifier = (readableIdentifier) => {
	const maxLength = 80;
	return readableIdentifier.length > maxLength ? readableIdentifier.substring(0, maxLength) + '...': readableIdentifier;
};
```
And then the result of every call to readableIdentifier is wrapped by the wrapReadableIdentifier  call, within the DefaultStatsFactoryPlugin.js.  So, as one example, this would change from:

```
                        if (error.module) {
				object.moduleIdentifier = error.module.identifier();
				object.moduleName = error.module.readableIdentifier(requestShortener);
			}
```

to this:

```
                        if (error.module) {
				object.moduleIdentifier = error.module.identifier();
				object.moduleName = wrapReadableIdentifier(error.module.readableIdentifier(requestShortener));
			}
```

Do I have it right?  On the surface, this seems to work.

**Comment by alexander-akait** [comment]
@trevorgithub Yes, it is right direction

**Comment by trevorgithub** [comment]
Thanks, @alexander-akait.  I made a pull request - https://github.com/webpack/webpack/pull/16479.  I need some advice on where I can add unit tests.  I couldn't find any existing ones for the area

**Comment by trevorgithub** [comment]
@alexander-akait, looks like some checks are failing on the build.  I ran the basic test locally, and it did fail for me as well.  However, the basic test also failed failed on the master branch, before my commit.
Do I need to sign the contributors licenses agreement to make a pull request?  (Even if a maintainer like yourself is ultimately merging?)

**Comment by alexander-akait** [comment]
We need update `test/StatsTestCases.basictest.js` 

> Do I need to sign the contributors licenses agreement to make a pull request? (Even if a maintainer like yourself is ultimately merging?)

yes, you need in any case

**Comment by trevorgithub** [comment]
I tried the update snapshots on the main branch, then re-ran yarn test:basic on that same main branch (i.e. without my changes).  That seemed to resolve the issues on the main branch.  Strange to me that git status shows clean after running the snapshot update.

**Comment by trevorgithub** [comment]
@alexander-akait , the update snapshots helped with some of the unit test problems.  But I've also changed the behaviour, limiting the scope of where the readable identifier is potentially truncated.  I only care about the external modules, and some of the tests were having problems with the broader shortening behaviour.  So, I changed the wrapper function to take the module, and only perform the shortening if it's an external module type, and the readable identifier is "too long".  Here's what the new function looks like:

```
/**
 * @param {string} readableIdentifier user readable identifier of the module
 * @param {Module} module base module type
 * @returns {string} an elided readableIdentifier, when readableIdentifier exceeds a maximum length
 */
const truncateLongExternalModuleReadableIdentifier = (
	readableIdentifier,
	module
) => {
	const maxLength = 80;
	return module instanceof ExternalModule &&
		readableIdentifier.length > maxLength
		? readableIdentifier.substring(0, maxLength) + "..."
		: readableIdentifier;
};
```

and here's an example usage:
```
                        if (error.module) {
				object.moduleIdentifier = error.module.identifier();
				object.moduleName = truncateLongExternalModuleReadableIdentifier(
					error.module.readableIdentifier(requestShortener),
					error.module
				);
			}
```

**Comment by ScriptedAlchemy** [comment]
You can use delegate modules "internal ./src/my-delegate.js"

**Comment by trevorgithub** [comment]
> You can use delegate modules "internal ./src/my-delegate.js"

Hi Zack,

Thanks for your reply.  I have no idea what it means though.  Can you please elaborate?

**Comment by trevorgithub** [comment]
@alexander-akait , thank you for your help on this PR.  I've looked around the web, and was able to determine what @ScriptedAlchemy referred to.  There's an example of using delegated modules in the module federation examples:
https://github.com/module-federation/module-federation-examples/tree/master/delegate-modules

By adopting this approach (I actually combined a variation of the code in [promised based dynamic remotes](https://webpack.js.org/concepts/module-federation/#promise-based-dynamic-remotes)) with the code in the above sample, I was able to achieve my desired functionality while also making the build output a lot cleaner.

As a result of this, I don't think my PR is needed, and it can be closed.
